src/Microsoft.AspNetCore.Mvc.Razor/RazorPage.cs

```diff
@@ -316,18 +314,7 @@ public static void WriteTo(TextWriter writer, HtmlEncoder encoder, object value)
             var htmlContent = value as IHtmlContent;
             if (htmlContent != null)
             {
-                var htmlTextWriter = writer as HtmlTextWriter;
-                if (htmlTextWriter == null)
-                {
-                    htmlContent.WriteTo(writer, encoder);
-                }
-                else
-                {
-                    // This special case allows us to keep buffering as IHtmlContent until we get to the 'final'
-                    // TextWriter.
-                    htmlTextWriter.Write(htmlContent);
-                }
-
```

Comment:
Considering bringing back `HtmlTextWriter` just to satisfy this special case. The other option is to always wrap the given writer in `ViewBufferTextWriter` inside Razor if it isn't one already.